scipy/optimize/_linprog.py

```diff
@@ -1072,12 +1345,77 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
     meth = method.lower()
     if options is None:
         options = {}
-
-    if meth == 'simplex':
-        return _linprog_simplex(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq,
-                                bounds=bounds, callback=callback, **options)
-    elif meth == 'interior-point':
-        return _linprog_ip(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq,
-                           bounds=bounds, callback=callback, **options)
-    else:
-        raise ValueError('Unknown solver %s' % method)
+    
+    # This needs to be before presolve for the tests to pass.
+    # If the problem is solved in presolve the error will not be raised
+    # leading the test to fail. 
+    # Should the check be moved to raise a ``NotImplementedError``
+    # only if callback would actually have been used?
```

Comment:
Sounds like a plan =).

With repsect to the format here's what the docs say:

```
callback : callable, optional

    If a callback function is provide, it will be called within each iteration of the simplex algorithm. The callback must have the signature callback(xk, **kwargs) where xk is the current solution vector and kwargs is a dictionary containing the following:

    "tableau" : The current Simplex algorithm tableau
    "nit" : The current iteration.
    "pivot" : The pivot (row, column) used for the next iteration.
    "phase" : Whether the algorithm is in Phase 1 or Phase 2.
    "basis" : The indices of the columns of the basic variables.
```

Most of these will still be relevant to the IP method. `phase` being the obvious exception. Maybe `kwargs` should be changed to be ` a dictionary containing callback arguments`?

Reply:
Actually, none of them are relevant to IP except `nit` and `xk`. But let's make this a separate PR.

Actually, regarding `xk` - is that the raw 'xk' that the algorithm is working with - with variables removed, added, and shifted according to bounds?

Come to think of it we might need feedback on this. Because aren't we changing what a callback function will see now that the simplex problem has gone through presolve? Could this break old code that uses callback?

Reply:
Even before these changes, these values corresponded with the standard form problem, not th user-specified problem with general bounds and inequality constraints. (Right?) The documentation doesn't explicitly spell out how the conversion was done or what the tableau means. The user would have had to reverse engineer all of that to make real user it it, right? For that reason I think it's reasonable that we be allowed to change it

When we add callback for IP, it should provide (only?) information that corresponds with the original problem, with all the standardization and presolving undone. Maybe just the current solution that corresponds with the original problem. The user can calculate constraint violations themselves. It doesn't make sense to expose the innards of the solver unless we plan to also publicly document the information needed to interpret it.